tests/ui/associated-inherent-types/dispatch-on-self-type-0.rs

```diff
@@ -0,0 +1,35 @@
+// check-pass
+
+#![feature(inherent_associated_types)]
+#![allow(incomplete_features)]
+
+// Check that inherent associated types are dispatched on the concrete Self type.
+
+struct Select<T>(T);
+
+impl Select<u8> {
+    type Projection = ();
+}
+
+impl Select<String> {
+    type Projection = bool;
+}
+
+struct Choose<T>(T);
+struct NonCopy;
+
+impl<T: Copy> Choose<T> {
+    type Result = Vec<T>;
+}
+
+impl Choose<NonCopy> {
+    type Result = ();
+}
+
+fn main() {
+    let _: Select<String>::Projection = false;
+    let _: Select<u8>::Projection = ();
+
+    let _: Choose<NonCopy>::Result = ();
+    let _: Choose<&str>::Result = vec!["..."];
```

Comment:
For followup work: it would be nice to see tests where we don't have a concrete type, but instead something from the environment.

Reply:
I've added two tests to make sure we use the correct `ParamEnv`. The first one is in fn `parameterized` in `dispatch-on-self-type-0.rs` which is `check-pass` and the second one is `not-found-unsatisfied-bounds-1.rs` which is `check-fail`. I did in fact not use the correct `ParamEnv` before. Now I use the environment of `block.owner` instead of the impl block. Let me know if you'd like to see more `ParamEnv` tests (in a follow-up PR maybe).